test/src/unit-class_const_iterator.cpp

```diff
@@ -220,20 +220,20 @@ TEST_CASE("const_iterator class")
             {
                 json j(json::value_t::null);
                 json::const_iterator it = j.cbegin();
-                CHECK(it.m_it.primitive_iterator == 1);
+                CHECK(it.m_it.primitive_iterator.m_it == 1);
```

Comment:
Well this a consequence of removing the `explicit` keyword on the new constructor.

As I mentioned a few days ago:

> In the second [example], the converting struct has template arguments, which prevents the sneaky implicit conversion, since the compiler cannot deduce T when trying to convert int to basic_json<T>.

Although code outside the `basic_json` class is protected from this *sneaky implicit conversion*, the code **inside** is not!

Every implicit conversion inside `basic_json` will try to use the `T &&` constructor, if it goes trough the enable_ifs.

It happened with `primitive_iterator` because there was an implicit conversion operator to `difference_type`, that's why I reimplemented the iterator operators (copy-pasted until I refactor, sorry about that).

That's also the reason why I made the `.m_it` suffix public, it was also needed in the tests, because Catch does fancy capture stuff, and seems to be able to call the doomed constructor (through a call to `operator<<(ostream&, primitive_iterator)` if I remember correctly)

I should have added comments explaining this, sorry about that too!